### PR TITLE
Support keeping MoE router projection unquantized for stability

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

Under quantization, MoE router gate computation was previously always quantized. During FP8 training, rounding errors on router logits can destabilize Top-K expert selection and impact convergence.

This PR adds an option to bypass router projection quantization via the `quantize_router_proj=false` configuration flag and Qwix rule interception.

BUGS: b/563540156

## Changes
* **Flag & Config:** `src/maxtext/configs/base.yml` and `src/maxtext/configs/types.py`
  - Added `quantize_router_proj` (default: `true` for backward compatibility). 
  - Added cross-validation ensuring `float32_gate_logits` and `quantize_router_proj` cannot both be `true`.
* **Qwix Interception:** `src/maxtext/layers/quantizations.py`
  - prepended `QtRule(module_path=r".*/gate$", weight_qtype=None, act_qtype=None, bwd_qtype=None, op_names=("dot_general",))` when `quantize_router_proj=false`.
  - Because Qwix matches rules in first-match order, this bypasses quantization for gate `dot_general` ops and falls back to `jax.lax.dot_general`.
* **Documentation:** `docs/reference/core_concepts/moe_configuration.md` and `docs/reference/core_concepts/quantization.md`
  - Documented flag behavior and training stability trade-offs 
  - Add clarifying note for other MoE flags related to training stability: `float32_gate_logits`, `float32_weight_sum`
* **Testing:** `tests/unit/quantizations_test.py`
  - Added unit test `RouterProjQwixInterceptionTest` 

### 1. Routing Mechanism and Stability
Router gating computes expert affinity scores via $s = \text{score\\_func}(x W_g) + b$, followed by Top-K ranking and score normalization. Pre-ranking scores operate within a narrow dynamic range; FP8 rounding errors on the projection $x W_g$ alter the relative ordering of borderline candidates. This triggers discrete routing flips, expert oscillation across training steps, and discontinuous gradient propagation across sparse expert branches.

### 2. Alignment with Industry Standards
* **DeepSeek-V3 ([paper](https://arxiv.org/abs/2412.19437), §3.3.1 "Mixed Precision Framework"):** FP8 is used for the compute-heavy linear-layer GEMMs (Fprop / Dgrad / Wgrad). The paper explicitly retains the original precision (`BF16`/`FP32`) for "the embedding module, the output head, **MoE gating modules**, normalization operators, and attention operators." MoE gating is excluded at module granularity.

* **Megatron-Core (`NVIDIA/Megatron-LM`):** The router weight is a plain `torch.nn.Parameter`, not a TransformerEngine `Linear`, so the gating projection is outside the TE FP8 path by construction. Precision is a separate knob: [`Router.gating`](https://github.com/NVIDIA/Megatron-LM/blob/3b1a08fc7e83c344653a4a54ae427c4f06a0d73d/megatron/core/transformer/moe/router.py#L114-L120) calls `router_gating_linear(input, self.weight, self.bias, router_dtype)` where `router_dtype` comes from [`moe_router_dtype`](https://github.com/NVIDIA/Megatron-LM/blob/3b1a08fc7e83c344653a4a54ae427c4f06a0d73d/megatron/core/transformer/transformer_config.py#L876-L879).

### 3. Compute and Parameter Footprint
For DeepSeek-V3 ($d_{\text{model}} = 7168$, $256$ routed experts, $58$ MoE layers):
* **Parameters:** Gate weight shape per layer is $[7168, 256] \approx 1.84\text{M}$ parameters ($3.67\text{ MB}$ in BF16). Across all 58 MoE layers (first 3 layers are dense), this represents $\approx 106\text{M}$ parameters total ($< 0.02\%$ of the 671B model).
* **FLOPs:** Gating GEMM consumes $< 0.5\%$ of total MoE layer compute per token ($3.68\text{M}$ FLOPs router vs. $792\text{M}$ FLOPs active experts).

Bypassing FP8 quantization on the router projection carries negligible memory and throughput overhead while resolving routing divergence.

# Tests

### Unit Tests

Tested Qwix interception tracing for `gate` `dot_general` ops under CPU abstract evaluation:
```bash
JAX_PLATFORMS=cpu pytest tests/unit/quantizations_test.py -k "RouterProjQwixInterceptionTest"
```

### Auxiliary Test

Verified Qwix interception behavior via standalone model tracing: [script](https://paste.googleplex.com/6568981897543680), [log](https://paste.googleplex.com/4549288265383936)
```
# quantize_router_proj=true (default: gate quantized)
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/MoeBlock_0/gate' op=dot_general0 rule=0
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/MoeBlock_0' op=einsum0 rule=None
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/shared_experts/wi_0' op=dot_general0 rule=0

# quantize_router_proj=false (gate unquantized)
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/MoeBlock_0/gate' op=dot_general0 rule=0
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/MoeBlock_0' op=einsum0 rule=None
[QWIX] module='decoder/moe_layers/DeepSeekMoeBlock_0/shared_experts/wi_0' op=dot_general0 rule=1
```

### E2E Validation

Evaluated on DeepSeek-V3 671B training runs with `quantization=fp8_full`(b/563540156)
- Profile: Confirmed MoE gate logits run unquantized in `bf16` (previously `fp8`).
- Pre-training Convergence: Steps to reach target loss is within 59 steps, on par with bf16 scheme.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by CaptainO5 — APPROVED** [review]

File: docs/reference/core_concepts/moe_configuration.md [hunk 1]
```diff
@@ -55,10 +55,15 @@ Dropping:
...
 
-`float32_weight_sum`: If enabled, performs the summation of expert weights using float32 precision for improved numerical stability. Recommended specifically when lower precision types cause convergence or quality issues.
+`float32_weight_sum`: Controls the accumulation precision of the MoE combine reduction — the weighted sum of the expert outputs $E_k(x)$ by their routing weights $g_k$ (not a sum of model parameters): $y = \sum_{k=1}^{K} g_k E_k(x)$
+
+- **`True` (default)**: Casts both operands to `float32` before the combine einsum, accumulates in `float32`, then casts back to the model `dtype`. Recommended for numerical stability.
+- **`False`**: Reduces directly in the model `dtype` (e.g. `bfloat16`), halving the operand size. Set `False` only for HBM-bound recipes if `bfloat16` accumulation does not degrade convergence.
 
 ### Routing Mechanism
 
+The router computes affinity logits via $s = \text{score\_func}(x W_g) + b$, where $x$ is token representations, $W_g$ is the gate projection kernel, and $b$ is optional routing bias.
```
Comment (RissyRan) [inline docs/reference/core_concepts/moe_configuration.md]: Could you help check the markdown? 

https://screenshot-v2.corp.google.com/5emnnvt4vhnio

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -217,9 +218,11 @@ float32_qk_product: false # in dot_product attention, whether to cast to fp32 th
 float32_logits: false # in dot_product attention, whether to cast to fp32 the inputs to softmax
 mla_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in the MLA Indexer (DeepSeek-V3 / DeepSeek-V3.2) across the unsharded local heads dimension natively.
 csa_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in the CSA Indexer (DeepSeek-V4) across the unsharded local heads dimension natively.
-float32_weight_sum: true # whether to use full fp32 precision to sum expert weights for numerical stability
-float32_gate_logits: false # whether to cast inputs to fp32 to compute MoE gate logits for numerical stability
-
+float32_weight_sum: true # whether to use fp32 accumulation for the MoE combine reduction (weighted sum of expert outputs by routing weights)
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml]: Nit: maybe add "after GMM"

File: src/maxtext/configs/base.yml [hunk 3]
```diff
@@ -217,9 +218,11 @@ float32_qk_product: false # in dot_product attention, whether to cast to fp32 th
...
+# whether to run the MoE gate (router) module in fp32 for numerical stability.
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml]: Maybe add comments after `float32_gate_logits` to distinguish with comment from `float32_weight_sum`?

File: src/maxtext/layers/quantizations.py [hunk 4]
```diff
@@ -958,7 +958,23 @@ def get_fp8_full_qwix_rule_w_sparsity(config: Config):
   else:
     module_path = "decoder/.*layers.*"
 
-  return [
+  # Qwix rules are evaluated in a first-match way
+  rules = []
+
+  if not config.quantize_router_proj:
```
Comment (RissyRan) [inline src/maxtext/layers/quantizations.py:980]: This part is very small, no obvious perf improvement I would assume?

File: src/maxtext/layers/quantizations.py [hunk 5]
```diff
@@ -969,15 +985,30 @@ def get_fp8_full_qwix_rule_w_sparsity(config: Config):
...
   """Returns a list of qwix.QtRule from `dtype`."""
 
   def make_qt_rule(dtype) -> list[qwix.QtRule]:
-    return [
+    # Qwix rules are evaluated in a first-match way
+    rules = []
+    if not config.quantize_router_proj:
+      # Setting all qtypes to None bypasses quantization, falling back to standard unquantized
+      # jax.lax.dot_general for both forward and backward passes.
+      rules.append(
```
Comment (RissyRan) [inline src/maxtext/layers/quantizations.py]: It seems duplicate with above. Could you combine this two together?

**Review by RissyRan — COMMENTED** [review]
LGTM, minor comments, thanks!

File: docs/reference/core_concepts/moe_configuration.md (hunk 1 above)
Comment (shuningjin) [reply docs/reference/core_concepts/moe_configuration.md]: Thanks for the catch!
- Updated from `$s = \text{score\_func}(x W_g) + b$` to `$s = \mathrm{score}(x W_g) + b$` (i.e., remove the underscore).
- Markdown preview is okay now:  $s = \mathrm{score}(x W_g) + b$. https://screenshot-v2.corp.google.com/1o4om521g90s8

File: src/maxtext/configs/base.yml (hunk 2 above)
Comment (shuningjin) [reply src/maxtext/configs/base.yml]: Thanks for the suggestion! Added "after GMM" for `float32_weight_sum`, and "before GMM" for `float32_gate_logits` to highlight distinction.

File: src/maxtext/configs/base.yml (hunk 3 above)
Comment (shuningjin) [reply src/maxtext/configs/base.yml]: I move both comments to the front for consistency. Placed "before" instead "after" for the long comment to be readable.
```
# float32_weight_sum - whether to use fp32 accumulation ...
float32_weight_sum: true
# float32_gate_logits - whether to run the MoE gate ...
float32_gate_logits: false
```

File: src/maxtext/layers/quantizations.py (hunk 5 above)
Comment (shuningjin) [reply src/maxtext/layers/quantizations.py]: Refactor with a helper function `_get_router_proj_unquantized_rule`, containing the rule and related comments.

File: src/maxtext/layers/quantizations.py (hunk 4 above)
Comment (shuningjin) [reply src/maxtext/layers/quantizations.py:980]: Yes, the TPS difference is very small for deepseek3-671b (e.g., 2).

**Review by RissyRan — APPROVED** [review]
